src/Compilers/CSharp/Portable/Binder/Semantics/Conversions/ConversionsBase.cs

```diff
@@ -1101,32 +1107,38 @@ private Conversion ClassifyImplicitBuiltInConversionFromExpression(BoundExpressi
             return Conversion.NoConversion;
         }
 
-        private Conversion ClassifyImplicitUtf8StringLiteralConversion(BoundExpression sourceExpression, CSharpCompilation compilation, TypeSymbol destination)
+#nullable enable
+        private Conversion ClassifyImplicitUtf8StringLiteralConversion(BoundExpression sourceExpression, TypeSymbol destination)
         {
+            var compilation = Compilation;
+
             // The language will allow conversions between string constants and byte sequences
             // where the text is converted into the equivalent UTF8 byte representation.
             // Specifically the compiler will allow for implicit conversions from string constants to
             // byte[], Span<byte>, and ReadOnlySpan<byte>.
-            ConstantValue constantValue = sourceExpression.ConstantValue;
+            ConstantValue? constantValue = sourceExpression.ConstantValue;
 
             TypeSymbol destinationOriginalDefinition = destination.OriginalDefinition;
             if (constantValue is ({ IsString: true } or { IsNull: true }) &&
                 sourceExpression.Type?.SpecialType == SpecialType.System_String &&
                 (destination is ArrayTypeSymbol { IsSZArray: true, ElementType.SpecialType: SpecialType.System_Byte } || // byte[]
                  (destinationOriginalDefinition.TypeKind == TypeKind.Struct && destinationOriginalDefinition.IsRefLikeType &&
+                  compilation is not null &&
```

Comment:
>compilation is not null && [](http://example.com/codeflow?start=18&length=26)

Could we simplify by asserting `Debug.Assert(compilation is not null);` at the start of the method? (Is there a caller that uses this with `Compilation == null` and doesn't hit the assert at the end of the method?)

Reply:
> Could we simplify by asserting ```Debug.Assert(compilation is not null);``` at the start of the method? (Is there a caller that uses this with ```Compilation == null``` and doesn't hit the assert at the end of the method?)

Technically it is possible to not reach the assert at the end even if the compilation is null. Given that the value comes from a property now, I am more comfortable with avoiding a possible crash. An assert is just as good as the tests that we have and will not protect us in a case of a test hole.